src/lib/core/modules/coderunner/runCode.js

```diff
@@ -1,28 +1,52 @@
-const vm = require('vm');
+const {NodeVM} = require('vm2');
 const fs = require('../../fs');
 
-const noop = function() {};
-
 class RunCode {
   constructor({logger}) {
     this.logger = logger;
-    const newGlobal = Object.create(global);
-    newGlobal.fs = fs;
-    this.context = Object.assign({}, {
-      global: newGlobal, console, exports, require, module, __filename, __dirname, process,
-      setTimeout, setInterval, clearTimeout, clearInterval
-    });
+    this.context = {env: require('../../env'), __dirname: fs.dappPath()};
+    
+    this.setupNodeVm();    
   }
 
-  doEval(code, tolerateError = false, forConsoleOnly = false) {
-    // Check if we want this code to run on the console or by user input. If it is by
-    // user input, we disallow `require` and `eval`.
-    let context = (forConsoleOnly) ? this.context : Object.assign({}, this.context, {
-      eval: noop, require: noop
+  setupNodeVm(){
+    this.vm = new NodeVM({
+      sandbox: this.context,
+      require: {
+        builtin: ["path", "env"],
+        mock: {
+          fs:{
+            access: fs.access,
+            diagramPath: fs.diagramPath,
+            dappPath: fs.dappPath,
+            embarkPath: fs.embarkPath,
+            existsSync: fs.existsSync,
+            ipcPath: fs.ipcPath,
+            pkgPath: fs.pkgPath,
+            readFile: fs.readFile,
+            readFileSync: fs.readFileSync,
+            readJSONSync: fs.readJSONSync,
+            readdir: fs.readdir,
+            readdirSync: fs.readdirSync,
+            stat: fs.stat,
+            statSync: fs.statSync,
+            tmpDir: fs.tmpDir
+          }
+        },
+        external: [
+          "@babel/runtime-corejs2/helpers/interopRequireDefault",
+          "@babel/runtime-corejs2/core-js/json/stringify",
+          "@babel/runtime-corejs2/core-js/promise",
+          "@babel/runtime-corejs2/core-js/object/assign",
+          "eth-ens-namehash"
+        ]
+      }
     });
+  }
 
+  doEval(code, tolerateError = false) {
     try {
-      return vm.runInNewContext(code, context);
+      return this.vm.run(code, code.indexOf('require') > -1 ? __filename : undefined);
```

Comment:
This parameter is a absolute path to load a require script (used in our EmbarkJS code generated script), however it seems that the only effect it has is in stack traces. I think this could be refined to figure which require it's looking for and actually resolve that filename explicitly, however in the case of EmbarkJS, we have several `require` statements, so I'm unsure how this would work.

From [the VM2 doco](https://github.com/patriksimek/vm2#loading-modules-by-relative-path):
> To load modules by relative path, you must pass full path of the script you're running as a second argument of vm's run method. Filename then also shows up in any stack traces produced from the script.
```
vm.run("require('foobar')", "/data/myvmscript.js");
```